File: python/ch26/simple_vis.py
# python/ch26/simple_vis.py
"""グリッド + 経路を ASCII / HTML で可視化（第26章）."""
from __future__ import annotations

import argparse
import html
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import List, Sequence, TextIO, Tuple

CELL_CHARS = {
    0: ".",
    1: "#",
    2: "S",
    3: "G",
}
PATH_CHAR = "*"


@dataclass
class VisInput:
    n: int
    grid: List[List[int]]
    path: List[Tuple[int, int]]


def read_vis(stream: TextIO) -> VisInput:
    lines = [ln.strip() for ln in stream.read().splitlines() if ln.strip()]
    n = int(lines[0])
    grid = [list(map(int, lines[i + 1].split())) for i in range(n)]
    path_len = int(lines[n + 1])
    path = []
    for i in range(path_len):
        r, c = map(int, lines[n + 2 + i].split())
        path.append((r, c))
    return VisInput(n=n, grid=grid, path=path)


def overlay_path(grid: Sequence[Sequence[int]], path: Sequence[Tuple[int, int]]) -> List[List[str]]:
    n = len(grid)
    view = [[CELL_CHARS.get(grid[r][c], "?") for c in range(n)] for r in range(n)]
    for r, c in path:
        if not (0 <= r < n and 0 <= c < n):
            continue
        if grid[r][c] in (2, 3):
            view[r][c] = CELL_CHARS[grid[r][c]]
        else:
            view[r][c] = PATH_CHAR
    return view


def render_ascii(view: Sequence[Sequence[str]]) -> str:
    rows = [" ".join(row) for row in view]
    border = "+" + "-" * (len(view[0]) * 2 - 1) + "+"
    body = "\n".join("| " + row + " |" for row in rows)
    return "\n".join([border, body, border])


def cell_color(ch: str) -> str:
    return {
        ".": "#f8f9fa",
        "#": "#343a40",
        "S": "#0d6efd",
        "G": "#198754",
        "*": "#ffc107",
    }.get(ch, "#dee2e6")


def render_html(view: Sequence[Sequence[str]], title: str) -> str:
    cells = []
    for row in view:
        for ch in row:
            bg = cell_color(ch)
            cells.append(
                f'<td style="width:28px;height:28px;text-align:center;'
                f'background:{bg};font-family:monospace;font-weight:bold;">'
                f"{html.escape(ch)}</td>"
            )
    table = "<tr>" + "</tr><tr>".join(
        "".join(cells[i : i + len(view[0])]) for i in range(0, len(cells), len(view[0]))
    ) + "</tr>"
    return f"""<!DOCTYPE html>
<html lang="ja">
<head><meta charset="utf-8"><title>{html.escape(title)}</title></head>
<body>
<h1>{html.escape(title)}</h1>
<p>凡例: . 空き / # 壁 / S 開始 / G ゴール / * 経路</p>
<table cellspacing="2" cellpadding="0">{table}</table>
</body>
</html>
"""


def default_data_path() -> Path:
    return Path(__file__).resolve().parents[2] / "data" / "ch26" / "sample.txt"


def main() -> None:
    parser = argparse.ArgumentParser(description="Grid path visualizer (ASCII / HTML)")
    parser.add_argument(
        "input",
        nargs="?",
        default=str(default_data_path()),
        help="input file (default: data/ch26/sample.txt)",
    )
    parser.add_argument("--html", action="store_true", help="emit HTML instead of ASCII")
    args = parser.parse_args()

    path = Path(args.input)
    with path.open(encoding="utf-8") as f:
        vis = read_vis(f)
    view = overlay_path(vis.grid, vis.path)
    title = f"ch26 path vis ({path.name}, n={vis.n}, steps={len(vis.path)})"

    if args.html:
        print(render_html(view, title))
    else:
        print(render_ascii(view))

    print(f"legend: . empty  # wall  S start  G goal  * path", file=sys.stderr)
    print(f"grid={vis.n} path_len={len(vis.path)}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: go/ch26/simple_vis.go
// go/ch26/simple_vis.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"html"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

var cellChars = map[int]string{
	0: ".",
	1: "#",
	2: "S",
	3: "G",
}

type VisInput struct {
	N    int
	Grid [][]int
	Path [][2]int
}

func readVis(path string) (VisInput, error) {
	f, err := os.Open(path)
	if err != nil {
		return VisInput{}, err
	}
	defer f.Close()

	sc := bufio.NewScanner(f)
	var lines []string
	for sc.Scan() {
		t := strings.TrimSpace(sc.Text())
		if t != "" {
			lines = append(lines, t)
		}
	}
	if err := sc.Err(); err != nil {
		return VisInput{}, err
	}
	n, _ := strconv.Atoi(lines[0])
	grid := make([][]int, n)
	for i := 0; i < n; i++ {
		grid[i] = parseInts(lines[i+1])
	}
	pathLen, _ := strconv.Atoi(lines[n+1])
	route := make([][2]int, pathLen)
	for i := 0; i < pathLen; i++ {
		p := parseInts(lines[n+2+i])
		route[i] = [2]int{p[0], p[1]}
	}
	return VisInput{n, grid, route}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func overlayPath(grid [][]int, path [][2]int) [][]string {
	n := len(grid)
	view := make([][]string, n)
	for r := 0; r < n; r++ {
		view[r] = make([]string, n)
		for c := 0; c < n; c++ {
			ch, ok := cellChars[grid[r][c]]
			if !ok {
				ch = "?"
			}
			view[r][c] = ch
		}
	}
	for _, p := range path {
		r, c := p[0], p[1]
		if r < 0 || c < 0 || r >= n || c >= n {
			continue
		}
		if grid[r][c] == 2 || grid[r][c] == 3 {
			view[r][c] = cellChars[grid[r][c]]
		} else {
			view[r][c] = "*"
		}
	}
	return view
}

func renderASCII(view [][]string) string {
	var b strings.Builder
	border := "+" + strings.Repeat("-", len(view[0])*2-1) + "+"
	b.WriteString(border + "\n")
	for _, row := range view {
		b.WriteString("| " + strings.Join(row, " ") + " |\n")
	}
	b.WriteString(border)
	return b.String()
}

func cellColor(ch string) string {
	switch ch {
	case ".":
		return "#f8f9fa"
	case "#":
		return "#343a40"
	case "S":
		return "#0d6efd"
	case "G":
		return "#198754"
	case "*":
		return "#ffc107"
	default:
		return "#dee2e6"
	}
}

func renderHTML(view [][]string, title string) string {
	var rows strings.Builder
	for _, row := range view {
		rows.WriteString("<tr>")
		for _, ch := range row {
			bg := cellColor(ch)
			rows.WriteString(fmt.Sprintf(
				`<td style="width:28px;height:28px;text-align:center;background:%s;font-family:monospace;font-weight:bold;">%s</td>`,
				bg, html.EscapeString(ch)))
		}
		rows.WriteString("</tr>")
	}
	return fmt.Sprintf(`<!DOCTYPE html>
<html lang="ja">
<head><meta charset="utf-8"><title>%s</title></head>
<body>
<h1>%s</h1>
<p>凡例: . 空き / # 壁 / S 開始 / G ゴール / * 経路</p>
<table cellspacing="2" cellpadding="0">%s</table>
</body>
</html>`, html.EscapeString(title), html.EscapeString(title), rows.String())
}

func main() {
	htmlOut := flag.Bool("html", false, "emit HTML instead of ASCII")
	input := flag.String("input", filepath.Join("data", "ch26", "sample.txt"), "input file")
	flag.Parse()

	vis, err := readVis(*input)
	if err != nil {
		fmt.Fprintf(os.Stderr, "read error: %v\n", err)
		os.Exit(1)
	}
	view := overlayPath(vis.Grid, vis.Path)
	title := fmt.Sprintf("ch26 path vis (%s, n=%d, steps=%d)", *input, vis.N, len(vis.Path))

	if *htmlOut {
		fmt.Println(renderHTML(view, title))
	} else {
		fmt.Println(renderASCII(view))
	}
	fmt.Fprintf(os.Stderr, "legend: . empty  # wall  S start  G goal  * path\n")
	fmt.Fprintf(os.Stderr, "grid=%d path_len=%d\n", vis.N, len(vis.Path))
}
